types/validator_set.go

```diff
@@ -74,35 +77,57 @@ func (vals *ValidatorSet) CopyIncrementProposerPriority(times int) *ValidatorSet
 // proposer. Panics if validator set is empty.
 // `times` must be positive.
 func (vals *ValidatorSet) IncrementProposerPriority(times int) {
+	if vals.IsNilOrEmpty() {
+		return
+	}
 	if times <= 0 {
 		panic("Cannot call IncrementProposerPriority with non-positive times")
 	}
 
 	// Cap the difference between priorities to be proportional to 2*totalPower by
 	// re-normalizing priorities, i.e., rescale all priorities by multiplying with:
 	//  2*totalVotingPower/(maxPriority - minPriority)
-	diffMax := 2 * vals.TotalVotingPower()
+	diffMax := K * vals.TotalVotingPower()
 	vals.RescalePriorities(diffMax)
+	vals.shiftByAvgProposerPriority()
 
 	var proposer *Validator
 	// call IncrementProposerPriority(1) times times:
 	for i := 0; i < times; i++ {
 		proposer = vals.incrementProposerPriority()
 	}
-	vals.shiftByAvgProposerPriority()
 
 	vals.Proposer = proposer
 }
 
+/*
+func (vals *ValidatorSet) RescalePriorities2() {
+	if vals.IsNilOrEmpty() {
+		return
+	}
+	tvp := vals.TotalVotingPower()
+	diff := computeMaxMinPriorityDiff(vals)
+	if diff > K * tvp {
+		ratio := float64(K * tvp) / float64(diff)
+		for _, val := range vals.Validators {
+			val.ProposerPriority = int64(float64(val.ProposerPriority) * ratio)
+		}
+	}
+}
+*/
+
 func (vals *ValidatorSet) RescalePriorities(diffMax int64) {
+	if vals.IsNilOrEmpty() {
+		return
+	}
 	// NOTE: This check is merely a sanity check which could be
 	// removed if all tests would init. voting power appropriately;
 	// i.e. diffMax should always be > 0
 	if diffMax <= 0 {
```

Comment:
related: https://github.com/tendermint/tendermint/pull/3049#discussion_r251341684

Reply:
Should we fix this in https://github.com/tendermint/tendermint/pull/3182?

Reply:
Yes, it's OK to this in  #3182 or in another PR.